File: android/app/src/test/java/com/dsview/player/UpdaterTest.kt
package com.dsview.player

import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class UpdaterTest {

    @Test
    fun `compareVersions maior patch vence`() {
        assertEquals(1, Updater.compareVersions("0.5.5", "0.5.4"))
    }

    @Test
    fun `compareVersions menor patch perde`() {
        assertEquals(-1, Updater.compareVersions("0.5.3", "0.5.4"))
    }

    @Test
    fun `compareVersions igual da zero`() {
        assertEquals(0, Updater.compareVersions("0.5.4", "0.5.4"))
    }

    @Test
    fun `compareVersions minor major pesam mais que patch`() {
        assertEquals(1, Updater.compareVersions("0.6.0", "0.5.99"))
        assertEquals(1, Updater.compareVersions("1.0.0", "0.99.99"))
    }

    @Test
    fun `findAndroidAsset acha o apk entre varios assets do release`() {
        val assets = JSONArray()
            .put(JSONObject().put("name", "dsview-windows-setup-0.4.3.exe").put("browser_download_url", "https://x/win-versioned.exe").put("size", 78000000))
            .put(JSONObject().put("name", "dsview-windows-setup.exe").put("browser_download_url", "https://x/win.exe").put("size", 78000000))
            .put(JSONObject().put("name", "dsview-android-0.5.5.apk").put("browser_download_url", "https://x/android-versioned.apk").put("size", 2900000))
            .put(JSONObject().put("name", "dsview-android.apk").put("browser_download_url", "https://x/android.apk").put("size", 2900000))
        val found = Updater.findAndroidAsset(assets)
        assertEquals("0.5.5", found?.version)
        assertEquals("https://x/android-versioned.apk", found?.url)
        assertEquals(2900000L, found?.size)
    }

    @Test
    fun `findAndroidAsset release sem apk do Android devolve null`() {
        val assets = JSONArray().put(JSONObject().put("name", "dsview-windows-setup-0.4.3.exe").put("browser_download_url", "https://x/a.exe").put("size", 1))
        assertNull(Updater.findAndroidAsset(assets))
    }

    @Test
    fun `findAndroidAsset lista vazia ou nula devolve null sem lancar`() {
        assertNull(Updater.findAndroidAsset(JSONArray()))
        assertNull(Updater.findAndroidAsset(null))
    }
}

File: windows/test/updater.test.js
'use strict';
require('./_stub-electron');
const test = require('node:test');
const assert = require('node:assert/strict');
const updater = require('../src/main/updater');

test('compareVersions: maior patch vence', () => {
  assert.equal(updater.compareVersions('0.4.3', '0.4.2'), 1);
});

test('compareVersions: menor patch perde', () => {
  assert.equal(updater.compareVersions('0.4.1', '0.4.2'), -1);
});

test('compareVersions: igual dá zero', () => {
  assert.equal(updater.compareVersions('0.4.2', '0.4.2'), 0);
});

test('compareVersions: minor/major pesam mais que patch', () => {
  assert.equal(updater.compareVersions('0.5.0', '0.4.99'), 1);
  assert.equal(updater.compareVersions('1.0.0', '0.99.99'), 1);
});

test('findWindowsAsset: acha o instalador do Windows entre vários assets do release', () => {
  const assets = [
    { name: 'dsview-android-0.5.5.apk', browser_download_url: 'https://x/android-versioned.apk', size: 100 },
    { name: 'dsview-android.apk', browser_download_url: 'https://x/android.apk', size: 100 },
    { name: 'dsview-windows-setup-0.4.3.exe', browser_download_url: 'https://x/win-versioned.exe', size: 78000000 },
    { name: 'dsview-windows-setup.exe', browser_download_url: 'https://x/win.exe', size: 78000000 },
  ];
  const found = updater.findWindowsAsset(assets);
  assert.equal(found.version, '0.4.3');
  assert.equal(found.url, 'https://x/win-versioned.exe');
  assert.equal(found.size, 78000000);
});

test('findWindowsAsset: release sem instalador do Windows devolve null', () => {
  const assets = [{ name: 'dsview-android-0.5.5.apk', browser_download_url: 'https://x/a.apk', size: 1 }];
  assert.equal(updater.findWindowsAsset(assets), null);
});

test('findWindowsAsset: lista vazia/undefined devolve null sem lançar', () => {
  assert.equal(updater.findWindowsAsset([]), null);
  assert.equal(updater.findWindowsAsset(undefined), null);
});
